apps/xcalc/man/xcalc.man
.\" t
.\" Copyright (c) 1994  X Consortium
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" Except as contained in this notice, the name of the X Consortium shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from the X Consortium.
.\"
.\"
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH XCALC 1 __xorgversion__
.SH NAME
xcalc \- scientific calculator for X
.SH SYNOPSIS
.nf
\fBxcalc\fP [\fB-stipple\fP] [\fB-rpn\fP] [-\fItoolkitoption...\fP]
\fBxcalc -version\fP
\fBxcalc -help\fP
.fi
.SH DESCRIPTION
.I xcalc
is a scientific calculator desktop accessory that can emulate a TI-30
or an HP-10C.
.SH OPTIONS
.PP
\fIxcalc\fP accepts all of the standard toolkit command line options along
with these additional options:
.PP
.TP 8
.B \-stipple
This option indicates that the background of the calculator should be
drawn using a stipple of the foreground and background colors.  On monochrome
displays improves the appearance.
.PP
.TP 8
.B \-rpn
This option indicates that Reverse Polish Notation should be used.  In this
mode the calculator will look and behave like an HP-10C.  Without this flag,
it will emulate a TI-30.
.TP 8
.B \-help
This option indicates that \fIxcalc\fP should print its usage message and exit.
.TP 8
.B \-version
This option indicates that \fIxcalc\fP should print its version and exit.
.SH OPERATION
.PP
.I Pointer Usage:
Operations may be performed with pointer button 1, or in some cases,
with the keyboard.
Many common calculator operations have keyboard accelerators.
To quit, press pointer button 3 on the AC key of the TI calculator,
or the ON key of the HP calculator.
.PP
.I Calculator Key Usage (TI mode):
The numbered keys, the +/- key, and the +, -, *, /, and = keys all do exactly
what you would expect them to.  It should be noted that the operators obey
the standard rules of precedence.  Thus, entering "3+4*5=" results in "23",
not "35".  The parentheses can be used to override this.  For example,
"(1+2+3)*(4+5+6)=" results in "6*15=90".
.PP
The entire number in the calculator display can be selected, in order to
paste the result of a calculation into text.
.PP
The action procedures associated with each function are given below.  These
are useful if you are interested in defining a custom calculator.
The action used for all digit keys is \fBdigit(\fIn\fP)\fR, where \fIn\fR
is the corresponding digit, 0..9.
.TP 10
.B 1/x
Replaces the number in the display with its reciprocal.
The corresponding action procedure is \fBreciprocal()\fR.
.TP 10
.B x^2
Squares the number in the display.
The corresponding action procedure is \fBsquare()\fR.
.TP 10
.B SQRT
Takes the square root of the number in the display.
The corresponding action procedure is \fBsquareRoot()\fR.
.TP 10
.B CE/C
When pressed once, clears the number in the display without clearing the state
of the machine.  Allows you to re-enter a number if you make a mistake.
Pressing it twice clears the state, also.
The corresponding action procedure for TI mode is \fBclear()\fR.
.TP 10
.B AC
Clears the display, the state, and the memory.  Pressing it with the third
pointer button turns off the calculator, in that it exits the program.
The action procedure to clear the state is \fBoff()\fR; to quit, \fBquit()\fR.
.TP 10
.B INV
Invert function.  See the individual function keys for details.
The corresponding action procedure is \fBinverse()\fR.
.TP 10
.B sin
Computes the sine of the number in the display, as interpreted by the current
DRG mode (see DRG, below).  If inverted, it computes the arcsine.
The corresponding action procedure is \fBsine()\fR.
.TP 10
.B cos
Computes the cosine, or arccosine when inverted.
The corresponding action procedure is \fBcosine()\fP.
.TP 10
.B tan
Computes the tangent, or arctangent when inverted.
The corresponding action procedure is \fBtangent()\fP.
.TP 10
.B DRG
Changes the DRG mode, as indicated by 'DEG', 'RAD', or 'GRAD' at the bottom
of the calculator ``liquid crystal'' display.
When in 'DEG' mode, numbers in the display are taken as being
degrees.  In 'RAD' mode, numbers are in radians, and in 'GRAD' mode, numbers
are in grads.  When inverted, the DRG key has a feature of
converting degrees to radians to grads and vice-versa.  Example:  put the
calculator into 'DEG' mode, and enter "45 INV DRG".  The display should now
show something along the lines of ".785398", which is 45 degrees converted to
radians.
The corresponding action procedure is \fBdegree()\fP.
.TP 10
.B e
The constant 'e'.  (2.7182818...).
The corresponding action procedure is \fBe()\fR.
.TP 10
.B EE
Used for entering exponential numbers.  For example, to get "-2.3E-4" you'd
enter "2 . 3 +/- EE 4 +/-".
The corresponding action procedure is \fBscientific()\fR.
.TP 10
.B log
Calculates the log (base 10) of the number in the display.  When inverted,
it raises "10.0" to the number in the display.
For example, entering "3 INV log" should result in "1000".
The corresponding action procedure is \fBlogarithm()\fP.
.TP 10
.B ln
Calculates the log (base e) of the number in the display.  When inverted,
it raises "e" to the number in the display.  For example, entering "e ln"
should result in "1".
The corresponding action procedure is \fBnaturalLog()\fR.
.TP 10
.B y^x
Raises the number on the left to the power of the number on the right.  For
example "2 y^x 3 =" results in "8", which is 2^3.  For a further example,
"(1+2+3) y^x (1+2) =" equals "6 y^x 3" which equals "216".
The corresponding action procedure is \fBpower()\fR.
.TP 10
.B not
Performs a bitwise not.
The corresponding action procedure is \fBnot()\fP.
.TP 10
.B and
Performs a bitwise and.
The corresponding action procedure is \fBand()\fP.
.TP 10
.B or
Performs a bitwise or.
The corresponding action procedure is \fBor()\fP.
.TP 10
.B xor
Performs a bitwise exclusive or.
The corresponding action procedure is \fBxor()\fP.
.TP 10
.B trunc
Truncates the number in the display to an integer.
The corresponding action procedure is \fBtrunc()\fP.
.TP 10
.B PI
The constant 'pi'.  (3.1415927....)
The corresponding action procedure is \fBpi()\fR.
.TP 10
.B x!
Computes the factorial of the number in the display.  The number in the display
must be an integer in the range 0-500, though, depending on your math library,
it might overflow long before that.
The corresponding action procedure is \fBfactorial()\fR.
.TP 10
.B (
Left parenthesis.  The corresponding action procedure for TI calculators
is \fBleftParen()\fR.
.TP 10
.B )
Right parenthesis.  The corresponding action procedure for TI calculators
is \fBrightParen()\fR.
.TP 10
.B base
Changes the number base, as indicated by 'DEC', 'HEX, or 'OCT' at the bottom
of the calculator display.
When in 'DEC' mode, numbers in the display are taken as being decimal
(base 10).  In 'HEX' mode, numbers are in hexadecimal (base 16), and in 'OCT'
mode, numbers are in octal (base 8).
The corresponding action procedure is \fBbase()\fP.
.TP 10
.B shl
Performs an arithmetic bitwise shift left,  For example, entering "1 shl 2"
should result in "4".
The corresponding action procedure is \fBshl()\fR.
.TP 10
.B shr
Performs an arithmetic bitwise shift right,  For example, entering "8 shr 1"
should result in "4".
The corresponding action procedure is \fBshr()\fR.
.TP 10
.B mod
Performs the modulo operation, which calculates the remainder when dividing
the first number by the second.  For example, entering "14 mod 8" should
result in "6".
The corresponding action procedure is \fBmod()\fR.
.TP 10
.B /
Division.  The corresponding action procedure is \fBdivide()\fR.
.TP 10
.B *
Multiplication.  The corresponding action procedure is \fBmultiply()\fR.
.TP 10
.B -
Subtraction.  The corresponding action procedure is \fBsubtract()\fR.
.TP 10
.B +
Addition.  The corresponding action procedure is \fBadd()\fR.
.TP 10
.B =
Perform calculation.  The TI-specific action procedure is \fBequal()\fR.
.TP 10
.B STO
Copies the number in the display to the memory location.
The corresponding action procedure is \fBstore()\fR.
.TP 10
.B RCL
Copies the number from the memory location to the display.
The corresponding action procedure is \fBrecall()\fR.
.TP 10
.B SUM
Adds the number in the display to the number in the memory location.
The corresponding action procedure is \fBsum()\fR.
.TP 10
.B EXC
Swaps the number in the display with the number in the memory location.
The corresponding action procedure for the TI calculator is \fBexchange()\fR.
.TP 10
.B +/-
Negate; change sign.  The corresponding action procedure is \fBnegate()\fR.
.TP 10
.B .
Decimal point.  The action procedure is \fBdecimal()\fP.
.sp
.PP
.I Calculator Key Usage (RPN mode):
The number keys, CHS (change sign), +, -, *, /, and ENTR keys all do exactly
what you would expect them to do.  Many of the remaining keys are the same as
in TI mode.  The differences are detailed below.  The action procedure
for the ENTR key is \fBenter()\fR.
.sp
.TP 10
.B <-
This is a backspace key that can be used if you make a mistake while entering
a number.  It will erase digits from the display.  (See BUGS).
Inverse backspace will clear the X register.
The corresponding action procedure is \fBback()\fR.
.TP 10
.B ON
Clears the display, the state, and the memory.  Pressing it with the third
pointer button turns off the calculator, in that it exits the program.
To clear state, the action procedure is \fBoff\fR; to quit, \fBquit()\fR.
.TP 10
.B INV
Inverts the meaning of the function keys.  This would be the  \fIf\fR key
on an HP calculator, but \fIxcalc\fR does not display
multiple legends on each key.  See the individual function keys for details.
.TP 10
.B 10^x
Raises "10.0" to the number in the top of the stack.
When inverted, it calculates the log (base 10) of the number in the display.
The corresponding action procedure is \fBtenpower()\fR.
.TP 10
.B e^x
Raises "e" to the number in the top of the stack.
When inverted, it calculates the log (base e) of the number in the display.
The action procedure is \fBepower()\fR.
.TP 10
.B STO
Copies the number in the top of the stack to a memory location.  There are 10
memory locations.  The desired memory is specified by following this
key with a digit key.
.TP 10
.B RCL
Pushes the number from the specified memory location onto the stack.
.TP 10
.B SUM
Adds the number on top of the stack to the number in the specified
memory location.
.TP 10
.B x:y
Exchanges the numbers in the top two stack positions, the X and Y registers.
The corresponding action procedure is \fBXexchangeY()\fR.
.TP 10
.B R v
Rolls the stack downward.  When inverted, it rolls the stack upward.
The corresponding action procedure is \fBroll()\fR.
.TP 10
.I blank
These keys were used for programming functions on the HP-10C.  Their
functionality has not been duplicated in \fIxcalc\fR.
.PP
Finally, there are two additional action procedures:
\fBbell()\fR, which rings the bell;
and \fBselection()\fR, which performs a cut on the
entire number in the calculator's ``liquid crystal'' display.
.SH ACCELERATORS
.PP
Accelerators are shortcuts for entering commands.  \fIxcalc\fP provides
some sample keyboard accelerators; also users can customize accelerators.
The numeric keypad accelerators provided by \fIxcalc\fP
should be intuitively correct.
The accelerators defined by \fIxcalc\fP on the main keyboard are given below:
.PP
.TS
l l l l l.
TI Key	HP Key	Keyboard Accelerator	TI Function	HP Function
_
SQRT	SQRT	r	squareRoot()	squareRoot()
AC	ON	space	clear()	clear()
AC	<-	Delete	clear()	back()
AC	<-	Backspace	clear()	back()
AC	<-	Control-H	clear()	back()
AC		Clear	clear()
AC	ON	q	quit()	quit()
AC	ON	Control-C	quit()	quit()

INV	i	i	inverse()	inverse()
sin	s	s	sine()	sine()
cos	c	c	cosine()	cosine()
tan	t	t	tangent()	tangent()
DRG	DRG	d	degree()	degree()

e		e	e()
ln	ln	l	naturalLog()	naturalLog()
y^x	y^x	^	power()	power()

PI	PI	p	pi()	pi()
x!	x!	!	factorial()	factorial()
(		(	leftParen()
)		)	rightParen()

/	/	/	divide()	divide()
*	*	*	multiply()	multiply()
-	-	-	subtract()	subtract()
+	+	+	add()	add()
\&=		\&=	equal()

0..9	0..9	0..9	digit()	digit()
.	.	.	decimal()	decimal()
+/-	CHS	n	negate()	negate()

	x:y	x		XexchangeY()
	ENTR	Return		enter()
	ENTR	Linefeed		enter()
.TE
.PP
.SH CUSTOMIZATION
The application class name is XCalc.
.PP
\fIxcalc\fP has an enormous application defaults file which specifies
the position, label, and function of each key on the calculator.
It also gives translations to serve as keyboard accelerators.
Because these resources are not specified in the source code, you can create
a customized calculator by writing a private application defaults file,
using the Athena Command and Form widget resources to specify the size and
position of buttons, the label for each button, and the function of
each button.
.PP
The foreground and background colors of each calculator key can be
individually specified.
For the TI calculator, a classical color resource specification might be:
.sp
.br
XCalc.ti.Command.background:	gray50
.br
XCalc.ti.Command.foreground:	white
.sp
For each of buttons 20, 25, 30, 35, and 40, specify:
.br
XCalc.ti.button20.background:	black
.br
XCalc.ti.button20.foreground:	white
.sp
For each of buttons 22, 23, 24, 27, 28, 29, 32, 33, 34, 37, 38, and 39:
.br
XCalc.ti.button22.background:	white
.br
XCalc.ti.button22.foreground:	black
.SH WIDGET HIERARCHY
In order to specify resources, it is useful to know the hierarchy of
the widgets which compose \fIxcalc\fR.  In the notation below,
indentation indicates hierarchical structure.  The widget class name
is given first, followed by the widget instance name.
.nf
.ft CW
XCalc xcalc
        Form  ti  \fIor\fR  hp    \fI(the name depends on the mode)\fR
                Form  bevel
                        Form  screen
                                Label  M
                                Toggle  LCD

libX11/NEWS
Version 1.1.2 - 2007-06-03

* Security fixes:
  * Fix CVE-2007-1667: Multiple integer overflows in the XGetPixel() and
    XInitImage() functions.
  * Bug #9279: Fixed a file descriptor leak.

* Xlib/XCB: Ensure _XReadEvents reads at least one new event and blocks for
  exactly one event.  Large performance improvement for XIfEvent and similar.

* Hold XCB's Xlib lock even when only the user lock (XLockDisplay) is held.
  An Xlib client can query Display state, such as with NextRequest, while it
  holds only the Xlib user lock (between XLockDisplay and XUnlockDisplay), so
  XCB requests in other threads should be blocked when the Xlib user lock is
  held.  We acquire the lock even when XInitThreads was not called, so that
  pure XCB code can use multiple threads even in an otherwise single-threaded
  Xlib application.

* Allow re-entrant Xlib calls from _XIOError.
  Some libraries try to clean up X resources from atexit handlers, _fini, or
  C++ destructors. To make these work, downgrade the Display lock to a user
  lock (as in XLockDisplay) before calling exit(3).  This blocks Xlib calls
  from threads other than the one calling exit(3) while still allowing the
  exit handlers to call Xlib.

* Xlib/XCB: Avoid re-crashing after _XIOError.

* Xlib/XCB: Only remove pending_requests when there are provably no more responses.

* For NetBSD, define XTHREADLIB and XTHREAD_CFLAGS.

* Bug #7703: Fixed XSetSizeHints() et al wrt use of uninitialized data.
  Now only those fields of the respective hint struct are set that
  are actually valid in the input data.

* Bug #10292: Fixed a memory leak related to XOpenDisplay() in the XCB code.

* Bug #7713: Initialize all of the event's fields before sending it.

* Bug #10562: Define _GNU_SOURCE on glibc systems.

* Use unistd.h to get getresuid() and friends, now that we have _GNU_SOURCE
  defined.

* Switched function definitions from K&R to ANSI style.

* Many constness fixes.

* Fixed a few warnings.

* Fix 64bit issues with reallocation.

* Manpage fixes:
  * Add man pages for XKB APIs.
    (Volunteer needed to convert prototypes in man pages to ANSI C style...)
  * Protect C comments and #defines in XKB man pages from being mangled by cpp
  * Fix typo in nroff macro in XkbAddGeomOverlayKey.man
  * Bug #9695: Fixed a few argument types in the XOpenIM manpage.
  * Markup tweak for XOpenIM.
  * Bug #9696: refer to XDefineCursor() instead of XDefineCusor().
  * Bug #9697: Fixed documentation of XVisualInfo struct.  The "depth" member
    was said to be unsigned int, but it's signed.

* nls fixes:
  * Update el_GR.UTF-8/Compose.pre to match changes in xkeyboard-config cvs.
  * For nls/*.pre, allow people to comment lines by starting them with '##'.

* Build system fixes:
  * Add autogen.sh to EXTRA_DIST.
  * Makefile.am: make ChangeLog hook safer
  * Don't distribute the configure-generated XlibConf.h in tarballs.

Version 1.1.1 - 2006-11-30

The "xcompmgr is hard, let's go releasing" release

* Fix Bug #9154 and friends, to unbreak xcompmgr:
  Always process an event for _XReadEvents, even if an error occurs.
  Previously, process_responses (in the wait_for_first_event case called
  from _XReadEvents) considered any return from xcb_wait_for_event
  sufficient to think it had processed an event. If xcb_wait_for_event
  returned an error, and no more events occurred before process_responses
  called xcb_poll_for_event, process_responses would try to return with
  dpy->head NULL, and would fail an assertion for the _XReadEvents
  postcondition. Now, process_responses continues using xcb_wait_for_event
  until it gets an event.

* Bug #9153: Fix access to freed memory.
  The fix for bug #8622 introduced a smaller bug where _XReply would
  read memory shortly after freeing it. This patch caches the needed
  value in a stack-allocated variable before the heap-allocated memory
  is freed.

* libx11 doesn't use inputproto in public headers; don't require it in x11.pc.
  Based on a Debian patch.

* Debian bug #354315: Clarify return value in XGetWindowAttributes man page.
  Patch by Debian user Ross Combs.


Version 1.1 - 2006-11-23

* Add note in man-page that XListFontsWithInfo is not thread-safe.  _XReply
  drops the Display lock, so the value of dpy->request may change before
  _XReply is called again.  Jamey Sharp discovered this by inspection a few
  years ago.

* Fix Bug #8622, by fixing the response processing order for threaded apps.
  process_responses (the common code for _XReply, _XReadEvents, and
  _XEventsQueued) now handles responses in order, by adding condition variables
  to the list of outstanding requests in dpy->xcb->pending_requests, and
  blocking on them when those requests should get processed, to allow _XReply
  to process them; if actually called from _XReply, it returns when _XReply's
  request should get processed.  _XReply broadcasts on its condition variable
  after it has read its reply and re-acquired the display lock.

* Don't hold the display lock around callbacks to the application. This avoids
  recursive locking of the display lock (which triggers an XCB locking
  assertion), particularly with emacs.

* Add xcb-xlib dependency to x11.pc when built against XCB.

* Allocate the right amount of memory for dpy->lock_fns.  Fixes a crash on
  startup with gdk.


Version 1.1 RC2 - 2006-11-02

Benno Schulenberg:
      nls: remove duplicate compose entries (bug #2286)
      nls: remove shadowed compose entries (bug #2286)
      nls (en_US): remove long compositions that override shorter (bug #2286)

Caolan McNamara:
      XKB geometry: fix leaks in GetKeyboardByName (bug #8099)

David Nusinow:
      Dynamically generate internal manpage section using __libmanpagesuffix__ so that it actually matches the section if you don't use 3X11

Jamey Sharp:
      Add correct Display locking to XKB functions.
      XKB bugfix: SyncHandle must be called after UnlockDisplay, not before.

apps/xcalc/xcalc.c
/*

Copyright (c) 1989  X Consortium

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of the X Consortium shall
not be used in advertising or otherwise to promote the sale, use or
other dealings in this Software without prior written authorization
from the X Consortium.

*/

/*
 * xcalc.c  -  a hand calculator for the X Window system
 *
 *  Original Author:  John H. Bradley, University of Pennsylvania
 *			(bradley@cis.upenn.edu)  March, 1987
 *  RPN mode added and port to X11 by Mark Rosenstein, MIT Project Athena
 *  Rewritten to be an Xaw and Xt client by Donna Converse, MIT X Consortium
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <signal.h>
#include <X11/Xatom.h>
#include <X11/Shell.h>
#include <X11/Xaw/Cardinals.h>
#include <X11/Xaw/Form.h>
#include <X11/Xaw/Label.h>
#include <X11/Xaw/Command.h>
#include <X11/Xaw/Toggle.h>
#include <X11/cursorfont.h>
#include "xcalc.h"

static Boolean convert(Widget w, Atom *selection, Atom *target, Atom *type,
		       XtPointer *value, unsigned long *length, int *format);
static void create_keypad(Widget parent);
static void create_display(Widget parent);
static void create_calculator(Widget shell);
static void done(Widget w, Atom *selection, Atom *target);
static void lose(Widget w, Atom *selection);
static void Syntax(int argc, char **argv);

/*
 *	global data
 */
int	rpn = 0;		/* Reverse Polish Notation (HP mode) flag */
char	dispstr[LCD_STR_LEN];		/* string to show up in the LCD */
Atom	wm_delete_window;		/* see ICCCM section 5.2.2 */

/*
 *	local data
 */
static Display	*dpy = NULL;		/* connection to the X server */
static Widget	toplevel=NULL;  	/* top level shell widget */
static Widget   calculator=NULL;	/* an underlying form widget */
static Widget	LCD = NULL;		/* liquid crystal display */
static Widget	ind[9];			/* mode indicators in the screen */
static char	selstr[LCD_STR_LEN]; /* storage for selections from the LCD */
					/* checkerboard used in mono mode */
static XtAppContext xtcontext;		/* Toolkit application context */
#define check_width 16
#define check_height 16
static unsigned char check_bits[] = {
   0x55, 0x55, 0xaa, 0xaa, 0x55, 0x55, 0xaa, 0xaa, 0x55, 0x55, 0xaa, 0xaa,
   0x55, 0x55, 0xaa, 0xaa, 0x55, 0x55, 0xaa, 0xaa, 0x55, 0x55, 0xaa, 0xaa,
   0x55, 0x55, 0xaa, 0xaa, 0x55, 0x55, 0xaa, 0xaa};

/*	command line options specific to the application */
static XrmOptionDescRec Options[] = {
{"-rpn",	"rpn",		XrmoptionNoArg,		(XtPointer)"on"},
{"-stipple",	"stipple",	XrmoptionNoArg,		(XtPointer)"on"}
};

/*	resources specific to the application */
static struct resources {
    Boolean	rpn;		/* reverse polish notation (HP mode) */
    Boolean	stipple;	/* background stipple */
    Cursor	cursor;
} appResources;

#define offset(field) XtOffsetOf(struct resources, field)
static XtResource Resources[] = {
{"rpn",		"Rpn",		XtRBoolean,	sizeof(Boolean),
     offset(rpn),	XtRImmediate,	(XtPointer) False},
{"stipple",	"Stipple",	XtRBoolean,	sizeof(Boolean),
     offset(stipple),	XtRImmediate,	(XtPointer) False},
{"cursor",	"Cursor",	XtRCursor,	sizeof(Cursor),
     offset(cursor),	XtRCursor,	(XtPointer)NULL}
};
#undef offset


int
main(int argc, char **argv)
{
    Arg		args[3];

    XtSetLanguageProc(NULL, (XtLanguageProc) NULL, NULL);

    /* Handle args that don't require opening a display */
    for (int n = 1; n < argc; n++) {
	const char *argn = argv[n];
	/* accept single or double dash for -help & -version */
	if (argn[0] == '-' && argn[1] == '-') {
	    argn++;
	}
	if (strcmp(argn, "-help") == 0) {
            Syntax(1, argv);
	}
	if (strcmp(argn, "-version") == 0) {
	    puts(PACKAGE_STRING);
	    exit(0);
	}
    }

    toplevel = XtAppInitialize(&xtcontext, "XCalc", Options, XtNumber(Options),
			       &argc, argv, NULL, NULL, 0);
    if (argc != 1) Syntax(argc, argv);

    XtSetArg(args[0], XtNinput, True);
    XtSetValues(toplevel, args, ONE);

    XtGetApplicationResources(toplevel, (XtPointer)&appResources, Resources,
			      XtNumber(Resources), (ArgList) NULL, ZERO);

    create_calculator(toplevel);

    XtAppAddActions(xtcontext, Actions, ActionsCount);

    XtOverrideTranslations(toplevel,
	   XtParseTranslationTable("<Message>WM_PROTOCOLS: quit()\n"));

    XtRealizeWidget(toplevel);

    dpy = XtDisplay(toplevel);
    wm_delete_window = XInternAtom(dpy, "WM_DELETE_WINDOW", False);
    XSetWMProtocols(dpy, XtWindow(toplevel), &wm_delete_window, 1);
    XDefineCursor(dpy, XtWindow(toplevel), appResources.cursor);

    if (appResources.stipple || (CellsOfScreen(XtScreen(toplevel)) <= 2))
    {
	Screen	*screen = XtScreen(toplevel);
	Pixmap	backgroundPix;

	backgroundPix = XCreatePixmapFromBitmapData
	    (dpy, XtWindow(toplevel),
	     (char *)check_bits, check_width, check_height,
	     WhitePixelOfScreen(screen), BlackPixelOfScreen(screen),
	     DefaultDepthOfScreen(screen));
	XtSetArg(args[0], XtNbackgroundPixmap, backgroundPix);
	XtSetValues(calculator, args, ONE);
    }

#ifndef IEEE
    signal(SIGFPE,fperr);
    signal(SIGILL,illerr);
#endif
    ResetCalc();
    XtAppMainLoop(xtcontext);

    return 0;
}

static void create_calculator(Widget shell)
{
    rpn = appResources.rpn;
    calculator = XtCreateManagedWidget(rpn ? "hp" : "ti", formWidgetClass,
				       shell, (ArgList) NULL, ZERO);
    create_display(calculator);
    create_keypad(calculator);
    XtSetKeyboardFocus(calculator, LCD);
}

/*
 *	Do the calculator data display widgets.
 */
static void create_display(Widget parent)
{
    Widget	bevel, screen;
    static Arg	args[] = {
    	{XtNborderWidth, (XtArgVal)0},
    	{XtNjustify, (XtArgVal)XtJustifyRight}
    };

    /* the frame surrounding the calculator display */
    bevel = XtCreateManagedWidget("bevel", formWidgetClass, parent,
				  (ArgList) NULL, ZERO);

    /* the screen of the calculator */
    screen = XtCreateManagedWidget("screen", formWidgetClass, bevel,
				   (ArgList) NULL, ZERO);

    /* M - the memory indicator */
    ind[XCalc_MEMORY] = XtCreateManagedWidget("M", labelWidgetClass, screen,
					args, XtNumber(args));

    /* liquid crystal display */
    LCD = XtCreateManagedWidget("LCD", toggleWidgetClass, screen, args,
				XtNumber(args));

    /* INV - the inverse function indicator */
    ind[XCalc_INVERSE] = XtCreateManagedWidget("INV", labelWidgetClass,
					 screen, args, XtNumber(args));

    /* DEG - the degrees switch indicator */
    ind[XCalc_DEGREE] = XtCreateManagedWidget("DEG", labelWidgetClass, screen,
					args, XtNumber(args));

    /* RAD - the radian switch indicator */
    ind[XCalc_RADIAN] = XtCreateManagedWidget("RAD", labelWidgetClass, screen,
					args, XtNumber(args));

    /* GRAD - the grad switch indicator */
    ind[XCalc_GRADAM] = XtCreateManagedWidget("GRAD", labelWidgetClass, screen,
					args, XtNumber(args));

    /* () - the parenthesis indicator */
    ind[XCalc_PAREN] = XtCreateManagedWidget("P", labelWidgetClass, screen,
					     args, XtNumber(args));

    /* HEX - the hexadecimal (base 16) indicator */
    ind[XCalc_HEX] = XtCreateManagedWidget("HEX", labelWidgetClass, screen,
					   args, XtNumber(args));

    /* DEC - the hexadecimal (base 16) indicator */
    ind[XCalc_DEC] = XtCreateManagedWidget("DEC", labelWidgetClass, screen,
					   args, XtNumber(args));

    /* OCT - the octal (base 8) indicator */
    ind[XCalc_OCT] = XtCreateManagedWidget("OCT", labelWidgetClass, screen,
					   args, XtNumber(args));
}

/*
 *	Do all the buttons.  The application defaults file will give the
 *	default button placement, default button labels, and default
 *      actions connected to the buttons.  The user can change any of
 *      these defaults in an environment-specific resource file.
 */

static void create_keypad(Widget parent)
{
    static const char	*Keyboard[] = {
	"button1", "button2", "button3", "button4", "button5",
	"button6", "button7", "button8", "button9", "button10",
	"button11","button12","button13","button14","button15",
	"button16","button17","button18","button19","button20",
	"button21","button22","button23","button24","button25",
	"button26","button27","button28","button29","button30",
	"button31","button32","button33","button34","button35",
	"button36","button37","button38","button39","button40",
	"button41","button42","button43","button44","button45",
	"button46","button47","button48","button49","button50",
	"button51","button52","button53","button54","button55",
	};
    register int i;
    int		 n = XtNumber(Keyboard);

    if (appResources.rpn) n--; 	/* HP has 39 buttons, TI has 40 */

    for (i=0; i < n; i++)
	XtCreateManagedWidget(Keyboard[i], commandWidgetClass, parent,
			      (ArgList) NULL, ZERO);
}

/*
 *	Miscellaneous utility routines that interact with the widgets.
 */

/*
 * 	called by math routines to write to the liquid crystal display.
 */
void draw(char *string)
{
    Arg	args[1];

    XtSetArg(args[0], XtNlabel, string);
    XtSetValues(LCD, args, ONE);
}
/*
 *	called by math routines to turn on and off the display indicators.
 */
void setflag(int indicator, Boolean on)
{
    if (on) XtMapWidget(ind[indicator]);
    else XtUnmapWidget(ind[indicator]);
}

/*
 *	ring the bell.
 */
void ringbell(void)
{
    XBell(dpy, 0);
}

/*
 *	die.
 */
void Quit(void)
{
    XtDestroyApplicationContext(xtcontext);
    exit(0);
}

/*
 *	recite and die.
 */
static void Syntax(int argc, char **argv)
{
    if (argc > 1) {
        fprintf(stderr, "%s: unknown options:", argv[0]);
        for (int i = 1; i <argc; i++)
            fprintf(stderr, " %s", argv[i]);
        fprintf(stderr, "\n\n");
    }
    fprintf(stderr, "Usage:  %s", argv[0]);
    for (Cardinal i = 0; i < XtNumber(Options); i++)
	fprintf(stderr, " [%s]", Options[i].option);
    fprintf(stderr, "\n");
    fprintf(stderr, "        %s -help\n", argv[0]);
    fprintf(stderr, "        %s -version\n", argv[0]);
    if (xtcontext != NULL)
        XtDestroyApplicationContext(xtcontext);
    exit((argc > 1) ? 1 : 0);
}

/*
 * I use actions on the toggle widget to support selections.  This
 * means that the user may not do a partial selection of the number
 * displayed in the `liquid crystal display.'  Copying numbers into
 * the calculator is also not supported.  So all you can do is copy
 * the entire number from the calculator display.
 */

/*ARGSUSED*/
static Boolean convert(Widget w, Atom *selection, Atom *target, Atom *type,
		       XtPointer *value, unsigned long *length, int *format)
{
    if (*target == XA_STRING)
    {
	*type = XA_STRING;
	*length = strlen(dispstr);
	strncpy(selstr, dispstr, (int)(*length));
	*value = selstr;
	*format = 8;
	return True;
    }
    return False;
}

/*
 * called when xcalc loses ownership of the selection.
 */
/*ARGSUSED*/
static void lose(Widget w, Atom *selection)
{
    XawToggleUnsetCurrent(LCD);
}

/*
 * called when some other client got the selection.
 */
/*ARGSUSED*/
static void done(Widget w, Atom *selection, Atom *target)
{
    selstr[0] = '\0';
}

/*
 * called by the selection() action routine, in response to user action.
 */
void do_select(Time time)
{
    Boolean	state;
    Arg		args[1];

    XtSetArg(args[0], XtNstate, &state);
    XtGetValues(LCD, args, 1);

    if (state)
        XtOwnSelection(LCD, XA_PRIMARY, time, convert, lose, done);
    else
	selstr[0] = '\0';
}
